src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -267,24 +267,36 @@ private ComponentState AttachAndInitComponent(IComponent component, int parentCo
         /// <param name="componentId">The unique identifier for the component within the scope of this <see cref="Renderer"/>.</param>
         /// <param name="eventHandlerId">The <see cref="RenderTreeFrame.AttributeEventHandlerId"/> value from the original event attribute.</param>
         /// <param name="eventArgs">Arguments to be passed to the event handler.</param>
-        public void DispatchEvent(int componentId, int eventHandlerId, UIEventArgs eventArgs)
+        /// <returns>
+        /// A <see cref="Task"/> which will complete once all asynchronous processing related to the event
+        /// has completed.
+        /// </returns>
+        public Task DispatchEventAsync(int componentId, int eventHandlerId, UIEventArgs eventArgs)
         {
             EnsureSynchronizationContext();
 
             if (_eventBindings.TryGetValue(eventHandlerId, out var binding))
             {
                 // The event handler might request multiple renders in sequence. Capture them
                 // all in a single batch.
+                var componentState = GetRequiredComponentState(componentId);
+                Task task = null;
                 try
                 {
                     _isBatchInProgress = true;
-                    GetRequiredComponentState(componentId).DispatchEvent(binding, eventArgs);
+                    task = componentState.DispatchEventAsync(binding, eventArgs);
+                }
+                catch (Exception ex) when (!task.IsCanceled)
+                {
+                    HandleException(ex);
```

Comment:
It took me a few minutes to realise what scenario L291 is for. Would something like the following make the flow more obvious, or am I misunderstanding the intended logic?

```cs
catch (Exception synchronousException)
{
    task = Task.FromException(synchronousException);
}
```

... so that the flow always continues to L299, and we don't need any extra special case for `IsCanceled`?

Reply:
Actually it probably does still need special handling for `IsCanceled` regardless.

Reply:
I based this on @pranavkm 's pr so I don't want to super-diverge from what's going on there. If we think this needs to change that's fine. 

Would comments make a big difference here? I don't relish the thought of allocating an extra task just for funsies.

Reply:
Sure, a comment would be totally sufficient.